doc/src/build/package-upgrades.md

```diff
@@ -138,6 +138,96 @@ module example::counter {
 }
 ```
 
+To upgrade a module using this pattern requires making two extra changes, on top of any implementation changes your upgrade requires:
+
+1. Bump the `VERSION` of the package
+2. Introduce a `migrate` function to upgrade the shared object:
+
+The following module is an upgraded `counter` that emits `Progress` events as originally discussed, but also provides tools for an admin
+(`AdminCap` holder) to prevent accesses to the counter from older package versions:
+
+```rust
+module example::counter {
+    use sui::event;
+    use sui::object::{Self, ID, UID};
+    use sui::transfer;
+    use sui::tx_context::{Self, TxContext};
+    // 1. Bump the `VERSION` of the package.
+    const VERSION: u64 = 2;
+    struct Counter has key {
+        id: UID,
+        version: u64,
+        admin: ID,
+        value: u64,
+    }
+    struct AdminCap has key {
+        id: UID,
+    }
+    struct Progress has copy, drop {
+        reached: u64,
+    }
+    /// Not the right admin for this counter
+    const ENotAdmin: u64 = 0;
+    /// Migration is not an upgrade
+    const ENotUpgrade: u64 = 1;
+    /// Calling functions from the wrong package version
+    const EWrongVersion: u64 = 2;
+    fun init(ctx: &mut TxContext) {
+        let admin = AdminCap {
+            id: object::new(ctx),
+        };
+        transfer::share_object(Counter {
+            id: object::new(ctx),
+            version: VERSION,
+            admin: object::id(&admin),
+            value: 0,
+        });
+        transfer::transfer(admin, tx_context::sender(ctx));
+    }
+    public entry fun increment(c: &mut Counter) {
+        assert!(c.version == VERSION, EWrongVersion);
+        c.value = c.value + 1;
+        if (c.value % 100 == 0) {
+            event::emit(Progress { reached: c.value })
+        }
+    }
+    // 2. Introduce a migrate function
+    entry fun migrate(c: &mut Counter, a: &AdminCap) {
+        assert!(c.admin == object::id(a), ENotAdmin);
+        assert!(c.version < VERSION, ENotUpgrade);
+        c.version = VERSION;
+    }
+}
+```
+
+Upgrading to this version of the package requires performing the package upgrade, and calling the `migrate` function in a follow-up
+transaction.  Note that the `migrate` function:
+
+- Is an `entry` function and **not `public`**.  This allows it to be entirely changed (including changing its signature or removing it
+  entirely) in later upgrades.
+- Accepts an `AdminCap` and checks that its ID matches the ID of the counter being migrated, making it a privileged operation.
+- Includes a sanity check that the version of the module is actually an upgrade for the object. This helps catch errors such as failing to bump the module version before upgrading.
+
+After a successful upgrade, calls to `increment` on the previous version of the package aborts on the version check, while calls on
+the later version should succeed.
+
+### Extensions
```

Comment:
Does this being a lower level heading to Upgrade Considerations make sense to you, @amnn ? This one and the one before. Or do you think it'd be better to be same level?

Reply:
This looks reasonable to me-- I don't think we want it to be a `##`

Reply:
Lower level makes sense to me too!